examples/shader/compute.rs

```diff
@@ -0,0 +1,115 @@
+use bevy::prelude::*;
+use bevy_render::{
+    dispatch::Dispatch,
+    pipeline::{
+        ComputePipeline, ComputePipelineDescriptor, ComputePipelineSpecialization, ComputePipelines,
+    },
+    render_graph::{base::node::COMPUTE_PASS, AssetRenderResourcesNode, RenderGraph},
+    renderer::RenderResources,
+    shader::{ComputeShaderStages, ShaderStage},
+};
+
+fn main() {
+    App::build()
+        .add_default_plugins()
+        .add_asset::<PrimeIndices>()
+        .add_startup_system(setup.system())
+        .run();
+}
+
+const COMPUTE_SHADER: &str = r#"
+#version 450
+layout(local_size_x = 1) in;
+
+layout(set = 0, binding = 0) buffer PrimeIndices_indices {
+    uint[] indices;
+}; // this is used as both input and output for convenience
+
+// The Collatz Conjecture states that for any integer n:
+// If n is even, n = n/2
+// If n is odd, n = 3n+1
+// And repeat this process for each new n, you will always eventually reach 1.
+// Though the conjecture has not been proven, no counterexample has ever been found.
+// This function returns how many times this recurrence needs to be applied to reach 1.
+uint collatz_iterations(uint n) {
+    uint i = 0;
+    if (n == 0) {
+        return n;
+    }
+    while(n != 1) {
+        if (mod(n, 2) == 0) {
+            n = n / 2;
+        }
+        else {
+            n = (3 * n) + 1;
+        }
+        i++;
+    }
+    return i;
+}
+
+void main() {
+    uint index = gl_GlobalInvocationID.x;
+    indices[index] = collatz_iterations(indices[index]);
+}
+"#;
+#[derive(Default, RenderResources)]
+struct PrimeIndices {
+    #[render_resources(buffer)]
+    indices: Vec<u32>,
+}
+
+#[derive(Default, Bundle)]
+struct ComputeComponents {
+    compute_pipelines: ComputePipelines,
+    dispatch: Dispatch,
+}
+
+// Setup our compute pipeline and a "dispatch" component/entity that will dispatch the shader.
+fn setup(
```

Comment:
I wouldn't consider this example "complete" until we have outputs copied back to the cpu side / printed to the console. Can we either add that or create an issue to follow up?

Reply:
I wouldn't want to merge this until the example actually did something the user could see. 👍 I haven't had time lately to take a deep dive into retrieving data back from a GPU resource.